model/map-jpa/src/main/java/org/keycloak/models/map/jpa/client/entity/JpaClientEntity.java

```diff
@@ -0,0 +1,622 @@
+/*
+ * Copyright 2021 Red Hat, Inc. and/or its affiliates
+ * and other contributors as indicated by the @author tags.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.keycloak.models.map.jpa.client.entity;
+
+import java.io.Serializable;
+import java.util.Collection;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.Iterator;
+import java.util.LinkedList;
+import java.util.List;
+import java.util.Map;
+import java.util.Objects;
+import java.util.Set;
+import java.util.UUID;
+import java.util.stream.Collectors;
+import javax.persistence.Basic;
+import javax.persistence.CascadeType;
+import javax.persistence.Column;
+import javax.persistence.Entity;
+import javax.persistence.FetchType;
+import javax.persistence.Id;
+import javax.persistence.OneToMany;
+import javax.persistence.Table;
+import org.hibernate.annotations.Generated;
+import org.hibernate.annotations.GenerationTime;
+import org.hibernate.annotations.Type;
+import org.hibernate.annotations.TypeDef;
+import org.hibernate.annotations.TypeDefs;
+import org.keycloak.models.map.client.MapClientEntity.AbstractClientEntity;
+import org.keycloak.models.map.client.MapProtocolMapperEntity;
+import static org.keycloak.models.map.jpa.client.storage.JpaClientMapStorage.SUPPORTED_VERSION;
+import org.keycloak.models.map.jpa.hibernate.jsonb.JsonbType;
+
+@Entity
+@Table(name = "client")
+@TypeDefs({@TypeDef(name = "jsonb", typeClass = JsonbType.class)})
+public class JpaClientEntity extends AbstractClientEntity implements Serializable {
+
+    @Id
+    @Column
+    private UUID id;
+
+    @Type(type = "jsonb")
+    @Column(columnDefinition = "jsonb")
+    private final JpaClientMetadata metadata = new JpaClientMetadata();
```

Comment:
Perhaps a shared instance could be used here rather than new instance with every `JpaClientEntity`?
```suggestion
    private final JpaClientMetadata metadata = JpaClientMetadata.INSTANCE;
```
Or maybe even `null`?

Reply:
Unfortunately it cannot. With both shared instance and `null` there are `NPE`s.

Reply:
Why is there NPE with a shared instance?

Reply:
NPE is comming from keycloak code, during startup there are default clients created and when shared instance is used none clients are persisted in database so when keycloak tries to use previously created client it gets `null`. I can try to investigate it further, but I don't exactly understand why we want to use shared instance in the first place. 

When we use shared instance, then we'd override values from instance by another instance and it doesn't seems correct. Could you please elaborate why there should be shared instance used?

Reply:
As discussed offline, instantiation of `JpaClientMetadata` was moved to constructor.